lib/stats/stats-query.c

```diff
@@ -28,125 +28,124 @@
 
 #include <string.h>
 
-typedef struct _NamedStatsCounterItem
-{
-  StatsCluster *cluster;
-  gint index;
-} NamedStatsCounterItem;
 
 static GHashTable *counter_index;
-typedef gboolean (*StatsClusterCounterCb)(StatsCluster *sc, NamedStatsCounterItem *named_counter,
-                                          StatsFormatCb format_cb, gpointer user_data);
-
 
-static const gchar *
-_get_counter_name_by_index(gint type_index)
+static void
+_setup_filter_expression(const gchar *expr, gchar **key_str)
 {
-  const gchar *_counter_type[SC_TYPE_MAX+1] =
-  {
-    "dropped",
-    "processed",
-    "stored",
-    "suppressed",
-    "stamp",
-    "MAX"
-  };
-
-  if (type_index > SC_TYPE_MAX)
-    return "";
-
-  return _counter_type[type_index];
+  if (!expr || g_str_equal(expr, ""))
+    {
+      *key_str = g_strdup("*");
+    }
+  else
+    {
+      *key_str = g_strdup(expr);
+    }
 }
 
-static const gchar *
-_get_name_of_counter_item(NamedStatsCounterItem *item)
+static gchar *
+_construct_counter_item_name(StatsCluster *sc, gint type)
 {
-  return _get_counter_name_by_index(item->index);
+  GString *name = g_string_new("");
+
+  g_string_append(name, sc->query_key);
+  g_string_append(name, ".");
+  g_string_append(name, stats_cluster_get_type_name(sc, type));
+
+  return g_string_free(name, FALSE);
 }
 
 static void
-_append_live_counters(StatsCluster *sc, gchar *ctr_str, GList **counters)
+_add_counter_to_index(StatsCluster *sc, gint type, StatsCounterItem *counter)
 {
-  gint i;
-  gboolean is_wildcard = FALSE;
-  if (ctr_str[0] == '*')
-    is_wildcard = TRUE;
+  gchar *counter_full_name = NULL;
 
-  for (i = 0; i < SC_TYPE_MAX; i++)
-    {
-      if (stats_cluster_is_alive(sc, i))
-        {
-          if (is_wildcard || !strcmp(_get_counter_name_by_index(i), ctr_str))
-            {
-              NamedStatsCounterItem *c = g_new0(NamedStatsCounterItem, 1);
-              c->cluster = sc;
-              c->index = i;
-              *counters = g_list_append(*counters, c);
-            }
-        }
-    }
+  counter_full_name = stats_counter_get_name(counter);
+  if (counter_full_name == NULL)
+    counter_full_name = _construct_counter_item_name(sc, type);
+
+  sc->counter_group.counters[type].name = counter_full_name;
+  g_hash_table_insert(counter_index, counter_full_name, &sc->counter_group.counters[type]);
+  sc->indexed_mask |= type;
```

Comment:
Factor out the counter registration to a separate function (and do not pass the `StatsCounterItem`  to that function).